### PR TITLE
fix(lsp): stop re-waking deleted-on-disk buffers; route the last element-ref splits through their owner (#1624, #1606)

### LINKED ISSUE #1606: compiler: element-ref split predicates duplicated beside the `split_element_ref` owner (values.rs `split_array_ref`/`is_array_ref`, var_command.rs `is_array_elem`, naming.rs `split_array_name_braced`)
Found by an independent centralisation audit of PR #1591's owners (verified at `claude/e4-codecs-subst` d5c369ec1). Drift-risk cleanup, **no observable divergence today**: `ends_with(')') && contains('(')` is logically identical to the owner's rule (one byte cannot satisfy both char tests), so these cannot diverge on any input — until one side changes.

#1591 established `tcl_syntax::naming::split_element_ref{,_bytes}` as the owner for the `name(index)` element split (governing C: `TclObjLookupVarEx`, `tmp/tcl9.0.4/generic/tclVar.c:594`, parse at `:683-686` — zero-length array name accepted). Remaining duplicates outside that PR's routing:

- `rust/tcl-compiler/src/codegen/values.rs:141` `split_array_ref` / `:152` `is_array_ref` — a full re-implementation of the split. The contract's exceptions list names `values.rs::is_bare_var_name` but not these.
- `rust/tcl-compiler/src/analyser/diagnostics/var_command.rs:2630` `is_array_elem`
- `rust/tcl-syntax/src/naming.rs:537` `split_array_name_braced` — re-derives the split inline **inside the owner's own file** rather than calling `split_element_ref`.

(The two VM copies the audit also found — `tcl-vm/src/command.rs:1456` and `tcl-vm/src/exec.rs:2254` — are being routed in-PR on #1591 since that PR already edits those files.)

**Deliberately out of scope** (correctly distinct, do not conflate): `sccp.rs:978 array_element_base` is a narrower fold-safety predicate with a rationale comment, and it correctly excludes the zero-length array name the owner admits.

**Fix shape:** route each through `split_element_ref` (or `_bytes`), delete the local copies, add the sites to the owner's consumer list in `docs/design/contracts/shared-utility-contracts-rust.md`, and mutation-verify (perturb the owner's guard, show a consumer test fails). Note `cargo xtask owner-resolution` validates the manifest only — it cannot detect a consumer that never calls the owner, which is why these survived.

Related: #1458 (the guard defect that motivated the owner), #1591.

### LINKED ISSUE #1624: LSP hardening: the #1619 cross-file wake set re-wakes a deleted-on-disk open buffer on every fact-moving publish
Follow-up hardening from #1619 / #1622. **Theoretical — no known user-visible impact**, and deliberately kept out of #1622 so that PR's before/after evidence stays a clean single-variable comparison.

## What

#1622 added a hand-built wake set to `publish_diagnostics_result`: when a publish moves a workspace fact, it also reschedules **open documents the workspace index does not hold**, because every other consumer set is read *from* the index and a just-opened document has no entry there yet.

The helper is `unindexed_open_documents`, taking the already-held `WorkspaceIndex` write guard and the `documents` map guard, returning the URIs to wake:

```rust
    docs.keys()
        .map(|uri| uri.as_str())
        .filter(|uri| !index.contains_document(uri))
        .map(ToOwned::to_owned)
        .collect()
```

"Open but unindexed" is normally a **transient** state — a document is in it only between its `didOpen` and its first debounced publish, after which `replace_document` puts it in the index and it leaves the set for good.

There is one way to be in it **permanently**: a buffer whose backing file was deleted out of band. `publish_diagnostics_result` deliberately calls `remove_document` for those rather than re-adding them (re-adding is what used to resurrect the ghost `did_change_watched_files` had just retired), so such a buffer is open and never indexed for as long as it stays open. Every fact-moving publish anywhere in the workspace then re-wakes it, and it re-analyses to reach the same answer.

## Why it is bounded, not a hang

Worth stating explicitly, because it is the reason this is a hardening and not a bug fix:

- The re-woken buffer's own publish has `before = absent, after = absent`, so its `change` is empty, `moved_workspace_facts()` is false, and **it wakes nobody**. There is no cascade and no cycle.
- `reschedule_peers` only bumps `generation` on a slot that is already `running`, so there is no unbounded worker spawn.
- Cost is therefore linear: at most one redundant re-analysis per deleted-backing-file open buffer per fact-moving publish.

It also cannot be the extension-host wedge in #1600 item 2: no `test-ext` fixture is ever deleted out of band, so the permanently-unindexed set is empty for the whole suite.

## Fix shape

Exclude orphaned buffers — the same `backing_file_deleted` flag `publish_diagnostics_result` already reads a few lines earlier (`lib.rs:3955`) to decide to remove rather than replace. Two lines, swapping `keys()` for `iter()` so the flag is reachable:

```rust
    docs.iter()
        .filter(|(_, doc)| !doc.backing_file_deleted)
        .map(|(uri, _)| uri.as_str())
        .filter(|uri| !index.contains_document(uri))
        .map(ToOwned::to_owned)
        .collect()
```

That leaves the set containing only *transiently* unindexed documents, which is what the #1619 fix actually targets, and makes it self-limiting by construction rather than by the empty-change argument above.

A regression test would need a document whose backing file is deleted while it stays open (the `did_change_watched_files` DELETED path already sets the flag), asserting it is not named a consumer by an unrelated document's fact-moving publish.

## Context

- #1619 — the race this wake set was added to close.
- #1622 — the PR that added it. The `did_open` half of that fix (keeping the index entry when the buffer matches disk) already shrinks this set considerably, since most opens no longer drop their entry at all.

*(Body edited once: the original included the full helper signature, and GitHub's sanitiser stripped the generic parameters inside the code fence. Signature described in prose instead.)*

---
Generated with [Claude Code](https://claude.com/claude-code)

### PR DESCRIPTION
**Pull request by bitwisecook**
## Summary

Two independent pre-release items, kept in separate commits, plus one review fix.

**#1624 — the cross-file wake set no longer re-wakes a deleted-on-disk buffer** (`rust/tcl-lsp-server/src/lib.rs`)

`unindexed_open_documents` (added by #1622 to close the #1619 didOpen race) wakes every open document the workspace index does not hold. That state is normally *transient* — a document is in it only between its `didOpen` and its first debounced publish — which is what makes the set self-limiting. A buffer whose backing file was deleted out of band is the one way to be in it permanently: the publish path deliberately calls `remove_document` for those rather than re-adding them, so every fact-moving publish anywhere in the workspace re-woke it. The `backing_file_deleted` flag now decides membership, so the set holds only transiently unindexed documents by construction.

**Review follow-up — an orphan stops *contributing* facts but keeps *consuming* them** (`da2c13463`)

Codex raised a P2 that the above conflated two roles one set was serving, and it was right. Absence from the index says two different things about an orphaned buffer:

- It must not **contribute**: its path is dead, so re-indexing duplicates every proc of the file it was renamed to and sends go-to-definition somewhere the editor cannot open. `remove_document` stays.
- It does still **consume**: the buffer is on screen and its squiggles are read. If it calls `helper 1` and a peer changes `helper`'s arity, its E002/W123 is wrong, and nothing could name it — the index-driven queries read a consumer's own call-site records *from* the index, and it has none there.

Reproduced first: with every pre-fix consumer set in play, the computed set for that scenario came back **empty**. The fix is a second, consumption-side set (`orphaned_fact_consumers`) rather than a weaker filter, so #1624's property survives — `unindexed_open_documents` stays the transient-race fallback and stays orphan-free. Targeted, not blanket: a source/package move wakes every orphan (it changes resolution without renaming anything, and is rare), while a signature change wakes only orphans whose text mentions the changed name — a conservative substring test, since any real call spells the name, so it cannot miss a consumer.

**#1606 — the last element-ref split duplicates route through their owner** (`rust/tcl-syntax`, `rust/tcl-compiler`, contract doc)

`tcl_syntax::naming::split_element_ref` owns the `name(index)` split (`TclObjLookupVarEx`, `tclVar.c(9.0.4):683-686`). Three sites still re-derived it:

- `codegen::values::split_array_ref` / `is_array_ref` — a full re-implementation. Now one-line facades over the owner, keeping their names so the ~18 codegen call sites are untouched; this matches the pattern #1591 used for the VM copies (`interp::elem_ref`, `command::looks_like_element`).
- `analyser::diagnostics::var_command`'s array-element constant harvest — a local `contains('(') && ends_with(')')` closure.
- `naming::split_array_name_braced_for_style` — re-derived inside the owner's own file.

No behaviour change: one byte cannot satisfy both char tests, so each copy decided every input exactly as the owner does. Each also carried the #1458 zero-length-array-name admission by luck rather than by contract.

`sccp::array_element_base` stays hand-rolled per the issue, and now says so at the site: it is a narrower fold-safety predicate whose non-empty-base test is the point.

The owner's contract entry gains an explicit consumer list (both VM sites, both compiler sites, both in-file sites), the deliberate `sccp` non-consumer, and a note that `cargo xtask owner-resolution` validates the manifest rather than call graphs — which is why these survived it.

### Mutation verification

Every fix was committed before mutating, and the whole battery re-run against the final tree.

| Mutation | Consumer test that fails |
| --- | --- |
| wake set admits orphans (`!doc.backing_file_deleted` relaxed) | `the_wake_set_skips_a_buffer_whose_backing_file_was_deleted` |
| consumption filter `\|\|` to `&&` (under-approximate — the review's defect) | `an_orphan_is_still_woken_by_a_signature_it_consumes`, `every_consumer_set_together_still_names_an_orphaned_caller` |
| consumption filter replaced by `true` (blanket re-wake — the #1624 waste) | the TN half of `an_orphan_is_still_woken_by_a_signature_it_consumes` |
| owner rejects the zero-length array name (#1458) | `codegen::values::tests::split_array_ref_matches_the_owners_edges` |
| owner refuses names containing `k` | `split_array_ref_basic` (facade), `w307_reads_the_constant_an_array_element_was_set_to` (harvest), `naming::split_array_name_braced` doctest (in-file split) — one per routed consumer |

The zero-length mutation found a real coverage gap first: it was invisible to the entire `tcl-compiler` suite (5898 tests green). Nothing at the consumer end pinned that half of the contract, and the `var_command` harvest had no coverage at all. Both now do.

### Considered and deliberately not done

- **Indexing an orphan "consumer-only"** — keeping its invocation records while dropping its definitions — would make the index-driven queries reach it directly. Rejected: `contains_document` / `document_uris` feed several production paths (whole-workspace consumer coverage, disk re-analysis), so putting a dead path back in that list risks the exact ghost `remove_document` exists to prevent, and it would move a `tcl-lsp-core` surface every consuming crate reads. The per-role sets stay in the one file that already decides the roles.
- **Three further element-split spellings outside #1606's audited list**, each carrying an extra condition so none is a pure duplicate: `analyser/diagnostics/usage.rs` (W216 `${arr($foo)}` shape, plus a non-empty-name test), `tcl-lsp-core/src/semantic_tokens.rs` (inline predicate), `compiler/value_shapes.rs` (plus a `!contains('}')` test). Noted on #1606 for a follow-up rather than folded in here.

## Validation

All commands inline-prefixed with `CARGO_INCREMENTAL=0 CARGO_PROFILE_DEV_DEBUG=0 CARGO_BUILD_JOBS=2`, run from `rust/`, crate-scoped (shared box). Re-run on the final tree after the review fix.

- `cargo fmt --all --check` — clean in both `rust/` and `runtime/rust/`
- `cargo clippy -p tcl-syntax --features num-bigint --all-targets` / `-p tcl-compiler --all-targets` / `-p tcl-lsp-server --all-targets` — no warnings
- `cargo check --workspace` — clean
- `cargo run -p xtask -- owner-resolution` — OK, 21 owner rows
- Consuming-crate suites, all green: `cargo test -p tcl-syntax --features num-bigint` (`num-bigint` is required — the `number_tower` test backend needs `num_traits`, a pre-existing condition on `rust`), `-p tcl-compiler` (5899 lib + all integration), `-p tcl-lsp-server` (478 lib, 1505 e2e, 22 preview-ticket, 14 smoke, 6 stdio), `-p tcl-lsp-core`, `-p tcl-vm`, `-p tcl-registry`, `-p tcl-spec-studio`

- [x] I ran relevant tests/checks locally and listed the exact commands.

## Compiler / diagnostics checklist

- [x] Did this change alter a compiler fact contract? No — the #1606 routing is behaviour-preserving, verified by equivalence and by the mutation table above.
- [x] The owning design doc was updated anyway: `docs/design/contracts/shared-utility-contracts-rust.md` gains the owner's consumer list and the `sccp` non-consumer note.
- [x] No diagnostic or optimisation code added or changed.
- [x] No new design doc.

Fixes #1624
Fixes #1606

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_01HqRAya5Xmdo4NJSWPm5C8o

### CODE REVIEW

File: rust/tcl-lsp-server/src/lib.rs [hunk 1]
```diff
@@ -3790,12 +3790,26 @@ impl IndexedDiagnosticChange {
...
+/// pure waste. The same `backing_file_deleted` flag the caller reads to choose
+/// removal over replacement decides it here.
 fn unindexed_open_documents(
     index: &core_workspace_index::WorkspaceIndex,
     docs: &HashMap<Uri, DocumentState>,
 ) -> Vec<String> {
-    docs.keys()
-        .map(|uri| uri.as_str())
+    docs.iter()
+        .filter(|(_, doc)| !doc.backing_file_deleted)
```
[the remark this answers is not in the document]
Comment (bitwisecook) [reply rust/tcl-lsp-server/src/lib.rs:3816]: Confirmed, and fixed in da2c13463 — you are right that this conflated index contribution with diagnostic consumption, and the union of every consumer set really did leave the orphan unreachable.

Reproduced your exact shape first, at the level where it bites. `live.tcl` changes `helper`'s arity; the orphan is on screen holding `helper 1` with a now-wrong E002. `command_diagnostic_consumers` reads a consumer's own call-site records *from* the index and the orphan has none there, and #1624 had just removed the one fallback that reached it, so the computed consumer set came back literally empty — the new test `every_consumer_set_together_still_names_an_orphaned_caller` fails with `got {}` against the pre-fix code.

The fix separates the roles rather than weakening the filter, because the two sets want opposite membership for the same buffer:

- Contribution, unchanged: `publish_diagnostics_result` still calls `remove_document` for an orphan. The dead path stays out of the index, so no resurrected ghost, no duplicate procs, no go-to-definition into a file that is gone.
- Consumption, new: `orphaned_fact_consumers` names orphans a fact-moving publish must still wake. `unindexed_open_documents` stays orphan-free, which keeps the #1624 property intact — that set is the transient-race fallback and it is self-limiting *only* because every document in it leaves at its first publish.

Targeted rather than blanket, so this is compatible with #1624 instead of a revert of it. A source or package move can change what the orphan's calls resolve to without renaming anything, so it is not filterable by name and wakes every orphan — those events are rare. A command-signature change wakes only orphans whose text mentions the changed name: a deliberately conservative substring test, since any real call spells the name, so it cannot miss a consumer, and the worst case is one redundant re-analysis for a name that appears only in a comment. The precise index rule needs exactly the call-site records this document does not have, so it is approximated on the safe side rather than re-derived.

Three tests now stand side by side and all pass together: the orphan is absent from the transient set, the orphan is woken by a signature it consumes and not by an unrelated name, and the union names it.

Mutation-verified in both directions, fix committed before mutating:

- `||` to `&&` in the consumption filter (under-approximate, your defect): `an_orphan_is_still_woken_by_a_signature_it_consumes` and `every_consumer_set_together_still_names_an_orphaned_caller` both fail.
- filter replaced by `true` (blanket re-wake, the #1624 waste): the TN half fails with "a name the orphan never mentions must not wake it".
- the original #1624 mutant re-run on the final tree still fails `the_wake_set_skips_a_buffer_whose_backing_file_was_deleted` only.

Gates re-run on the final tree: fmt clean, clippy `-p tcl-lsp-server --all-targets` clean, `cargo check --workspace` clean, full tcl-lsp-server suite green (478 lib, 1505 e2e, 22 preview-ticket, 14 smoke, 6 stdio).

🤖 Generated with [Claude Code](https://claude.com/claude-code)

---
_Generated by [Claude Code](https://claude.ai/code)_
